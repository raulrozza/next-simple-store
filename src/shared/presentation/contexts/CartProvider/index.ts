export { default as useAddToCart } from './hooks/useAddToCart';
