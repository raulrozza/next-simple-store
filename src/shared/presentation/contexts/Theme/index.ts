export { default as useTheme } from './hooks/useTheme';
export { default as Theme } from './implementations/StyledComponentTheme';
