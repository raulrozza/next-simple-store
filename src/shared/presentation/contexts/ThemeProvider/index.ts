export { default as useTheme } from './hooks/useTheme';
export { default as ThemeProvider } from './implementations/StyledComponentTheme';
