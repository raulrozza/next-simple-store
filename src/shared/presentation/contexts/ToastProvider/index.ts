export { useToastProvider } from './context';
export { default as ToastProvider } from './implementations/ReactToastifyProvider';
