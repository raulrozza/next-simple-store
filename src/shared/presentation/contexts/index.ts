export * from './CartProvider';
export * from './ThemeProvider';
