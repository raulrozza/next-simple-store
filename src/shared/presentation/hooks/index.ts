import { trpc } from '@/shared/infra/http/trpc';

const { useQuery, useMutation } = trpc;

export { useQuery, useMutation };
