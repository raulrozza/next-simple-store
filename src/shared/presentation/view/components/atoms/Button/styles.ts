import styled, { css } from 'styled-components';

interface StyledButtonProps {
    variant?: 'primary' | 'secondary';
}

const VARIANT_STYLES = {
    primary: css`
        ${({ theme }) => css`
            background-color: ${theme.palette.primary['700']};
            color: ${theme.palette.gray['0']};

            &:hover {
                background-color: ${theme.palette.primary['600']};
            }
        `}
    `,
    secondary: css`
        ${({ theme }) => css`
            background-color: ${theme.palette.secondary['700']};
            color: ${theme.palette.gray['0']};

            &:hover {
                background-color: ${theme.palette.secondary['600']};
            }
        `}
    `,
    default: css`
        ${({ theme }) => css`
            color: ${theme.palette.gray['900']};

            &:hover {
                background-color: ${theme.palette.gray['200']};
            }
        `}
    `,
};

export const StyledButton = styled.button<StyledButtonProps>`
    ${({ theme, variant }) => css`
        padding: ${theme.layout.spacing(2)};
        border-radius: ${theme.layout.borderRadius.md};
        transition: background-color 0.2s;

        display: flex;
        align-items: center;
        gap: ${theme.layout.spacing(1)};

        ${VARIANT_STYLES[variant || 'default']}
    `}
`;
