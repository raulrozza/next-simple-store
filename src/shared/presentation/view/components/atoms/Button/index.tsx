import { ButtonHTMLAttributes, FC } from 'react';

import { FaPlus } from 'react-icons/fa';

import { StyledButton } from './styles';

const ICONS = {
    plus: FaPlus,
};

interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
    variant?: 'primary' | 'secondary';
    icon?: keyof typeof ICONS;
    asAnchor?: boolean;
}

const Button: FC<ButtonProps> = ({ children, icon, asAnchor, ...props }) => {
    const Icon = icon && ICONS[icon];

    return (
        <StyledButton as={asAnchor ? 'a' : 'button'} {...props}>
            {Icon && <Icon data-testid="button-Icon" />}
            {children}
        </StyledButton>
    );
};

export default Button;
