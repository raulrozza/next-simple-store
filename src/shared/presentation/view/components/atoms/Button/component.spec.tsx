import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import theme from '@/config/theme';
import { ThemeProvider } from '@/shared/presentation/contexts';

import Button from './index';

describe('Button', () => {
    it('should render the component correctly', () => {
        const view = render(<Button>My button</Button>, {
            wrapper: ThemeProvider,
        });

        expect(screen.getByText('My button')).toBeInTheDocument();
        expect(screen.getByText('My button')).toBeInstanceOf(HTMLButtonElement);
        expect(screen.queryByTestId('button-Icon')).toBeNull();
        expect(view).toMatchSnapshot();
    });

    it('should dispath the on click function when clicking the button', async () => {
        const user = userEvent.setup();
        const onClick = jest.fn();

        render(
            <Button onClick={onClick} role="button">
                My button
            </Button>,
            {
                wrapper: ThemeProvider,
            },
        );

        await user.click(screen.getByRole('button'));

        expect(onClick).toHaveBeenCalled();
    });

    it('should render an icon if the icon prop is passed', () => {
        const view = render(<Button icon="plus">My button</Button>, {
            wrapper: ThemeProvider,
        });

        expect(screen.getByTestId('button-Icon')).toBeInTheDocument();
        expect(view).toMatchSnapshot();
    });

    it('should render the element as an anchor if the asAnchor prop is passed', () => {
        const view = render(<Button asAnchor>My button</Button>, {
            wrapper: ThemeProvider,
        });

        expect(screen.getByText('My button')).toBeInstanceOf(HTMLAnchorElement);
        expect(view).toMatchSnapshot();
    });

    describe('styles', () => {
        it('should render each button variant with the correct styles', () => {
            const view = render(
                <>
                    <Button variant="primary">Primary</Button>
                    <Button variant="secondary">Secondary</Button>
                    <Button>Default</Button>
                </>,
                {
                    wrapper: ThemeProvider,
                },
            );

            expect(screen.getByText('Primary')).toHaveStyle({
                backgroundColor: theme.palette.primary['700'],
                color: theme.palette.gray['0'],
            });
            expect(screen.getByText('Secondary')).toHaveStyle({
                backgroundColor: theme.palette.secondary['700'],
                color: theme.palette.gray['0'],
            });
            expect(screen.getByText('Default')).toHaveStyle({
                backgroundColor: undefined,
                color: theme.palette.gray['900'],
            });
            expect(view).toMatchSnapshot();
        });
    });
});
