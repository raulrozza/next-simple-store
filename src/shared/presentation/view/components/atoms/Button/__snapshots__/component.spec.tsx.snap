// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render an icon if the icon prop is passed 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  color: #1E4243;
}

.c0:hover {
  background-color: #CDD5D5;
}

<body>
    <div>
      <button
        class="c0"
      >
        <svg
          data-testid="button-Icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          />
        </svg>
        My button
      </button>
    </div>
  </body>,
  "container": .c0 {
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  color: #1E4243;
}

.c0:hover {
  background-color: #CDD5D5;
}

<div>
    <button
      class="c0"
    >
      <svg
        data-testid="button-Icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
        />
      </svg>
      My button
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Button should render the component correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  color: #1E4243;
}

.c0:hover {
  background-color: #CDD5D5;
}

<body>
    <div>
      <button
        class="c0"
      >
        My button
      </button>
    </div>
  </body>,
  "container": .c0 {
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  color: #1E4243;
}

.c0:hover {
  background-color: #CDD5D5;
}

<div>
    <button
      class="c0"
    >
      My button
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Button should render the element as an anchor if the asAnchor prop is passed 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  color: #1E4243;
}

.c0:hover {
  background-color: #CDD5D5;
}

<body>
    <div>
      <a
        class="c0"
      >
        My button
      </a>
    </div>
  </body>,
  "container": .c0 {
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  color: #1E4243;
}

.c0:hover {
  background-color: #CDD5D5;
}

<div>
    <a
      class="c0"
    >
      My button
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Button styles should render each button variant with the correct styles 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  background-color: #579FB3;
  color: #FFFFFF;
}

.c0:hover {
  background-color: #71B7C8;
}

.c1 {
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  background-color: #754A56;
  color: #FFFFFF;
}

.c1:hover {
  background-color: #845864;
}

.c2 {
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  color: #1E4243;
}

.c2:hover {
  background-color: #CDD5D5;
}

<body>
    <div>
      <button
        class="c0"
      >
        Primary
      </button>
      <button
        class="c1"
      >
        Secondary
      </button>
      <button
        class="c2"
      >
        Default
      </button>
    </div>
  </body>,
  "container": .c0 {
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  background-color: #579FB3;
  color: #FFFFFF;
}

.c0:hover {
  background-color: #71B7C8;
}

.c1 {
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  background-color: #754A56;
  color: #FFFFFF;
}

.c1:hover {
  background-color: #845864;
}

.c2 {
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  color: #1E4243;
}

.c2:hover {
  background-color: #CDD5D5;
}

<div>
    <button
      class="c0"
    >
      Primary
    </button>
    <button
      class="c1"
    >
      Secondary
    </button>
    <button
      class="c2"
    >
      Default
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
